Handle folder selection with external command

AckslD opened the pull request:
Closes #6218 

@The-Compiler folder selection is now implemented. I had some remaining questions (see also TODOs in the files):
1. Where does the value `2` actually come from regarding folder mode?
1. Should we remain internal variables/functions which refer to file(s) since they might now also refer to folders?
1. Should we assert that a file/folder is actually chosen when it is supposed to be?
1. Should the type of the `mode` argument to `webview.chooseFiles` be `int` instead of `QWebEnginePage.FileSelectionMode` since it can actually take other values?
1. Does anyone know if this feature works with `nnn`, I couldn't find it in the docs.
1. Is it possible to ignore the requirement for doc-strings? In some cases it might be unnecessary clutter.

The-Compiler commented on tests/end2end/data/fileselect.html:
```diff
@@ -17,5 +17,6 @@
     <body onload="onLoad()">
         <input type="file" id="single_file" name="single_file" accept=".txt">
         <input type="file" id="multiple_files" name="multiple_files" accept=".txt" multiple>
+        <input type="file" id="folder" name="folder" webkitdirectory>
```
Maybe add some text here (above the inputs) as well, so that the page is a bit easier to use for manual testing as well? Right now it's a bit confusing:

![image](https://user-images.githubusercontent.com/625793/110362855-1d239480-8042-11eb-8ca6-6227e3c8e8f1.png)

AckslD commented on qutebrowser/browser/shared.py:
```diff
@@ -421,8 +436,9 @@ def _execute_fileselect_command(
 
-    if not multiple:
+    if not qb_mode == FileSelectionMode.MULTIPLE_FILES:
+        # TODO should we assert if a file(s)/folder was actually chosen?
```
This is what I found:
* Providing a file path which doesn't exist gives the warning
   ```
   WARNING: Ignoring invalid item in FilePickerController::accepted(QStringList): non-existent
    ```
    and the file is correctly ignored. The same for single_file, multiple_files and folder.
* Selecting a directory path instead of a file or vice versa does get accepted in the test page. I also tried it here to select a file to attach to the message and instead choose a folder. This triggered the error from github:
  ```
   We don't support that file type. Try again with a GIF, JPEG, JPG, PNG, DOCX, GZ, LOG, PDF, PPTX, TXT, XLSX or ZIP
   ```
    (there were two .txt files in the folder)

AckslD commented:
@The-Compiler do you know why the archlinux-webkit test fails?

The-Compiler commented:
Looking at [the qutebrowser logs](https://github.com/qutebrowser/qutebrowser/pull/6233/checks?check_run_id=2065534908#step:5:342), it fails with:

```
  10:21:01.112 DEBUG    procs      guiprocess:_pre_start:159 Executing: xterm -e ranger --choosefile=/tmp/qutebrowser-fileselect-t6999inl
  10:21:01.120 ERROR    message    message:error:59 Error while spawning choose-file: execvp: No such file or directory
```

which makes sense, since you don't call `shared.choose_file(qb_mode=shared.FileSelectionMode.folder)` anywhere. Given that [QtWebKit doesn't seem to support `webkitdirectory`](https://github.com/qtwebkit/qtwebkit/blob/00698c586f552cb9cf753a8d060df855e694f2a4/Source/WebKitLegacy/qt/WidgetApi/qwebpage.cpp#L347-L368) at all (despite the name!), I think it's fine to just add `@qtwebkit_skip` above that test in the .feature file.

The [other failure](https://github.com/qutebrowser/qutebrowser/pull/6233/checks?check_run_id=2065534908#step:5:548) is unrelated, and just seems like [Hypothesis](https://hypothesis.readthedocs.io/en/latest/index.html) getting lucky and finding a case it never found before. Feel free to ignore it, I'll fix it in master.

Other than that, the only open question is [whether to validate files](https://github.com/qutebrowser/qutebrowser/pull/6233#discussion_r590169636) - I think it'd be a good idea to do so, so we can make this visible to the user (a warning in the terminal probably won't be seen). What do you think?

AckslD commented:
@The-Compiler Ah that makes sense, I stopped looking when I saw the failure caused by Hypothesis so didn't see the other one. I added the flag to skip that test.

Regarding validating files, yeah I think it makes sense to check if the files/folders exists and warn the use otherwise (I've added this + a test). Selecting a folder instead of a file I think we should just allow since this seems to be the behaviour anyway from what I can tell.

AckslD commented:
Strange! In the tests non-existent files are logged from the test page, which does not happen when I test manually...

The-Compiler commented:
I can reproduce when it's an absolute rather than relative path:

```
$ python3 -m qutebrowser --temp-basedir --debug --logfilter js -s fileselect.handler external -s fileselect.single_file.command '["echo", "/non-existent.txt"]' tests/end2end/data/fileselect.html  ':later 1000 click-element id single_file'
23:02:56 WARNING  message    message:warning:71 Selected non-existent file '/non-existent.txt'
23:02:56 DEBUG    js         shared:javascript_log_message:156 [file:///home/florian/proj/qutebrowser/git/tests/end2end/data/fileselect.html:11] Files: non-existent.txt
```

After [looking at the relevant code](https://github.com/qt/qtwebengine/blob/v5.15.2/src/core/file_picker_controller.cpp#L73-L77) I think your conclusion:

> * Providing a file path which doesn't exist gives the warning
>    ```
>    WARNING: Ignoring invalid item in FilePickerController::accepted(QStringList): non-existent
>     ```
>     and the file is correctly ignored. The same for single_file, multiple_files and folder.

isn't accurate - I assume you had provided a relative path while testing, but relative paths aren't supported at all according to that code. With absolute paths, it looks like QtWebEngine doesn't do any validation whether the file exists or not.

Thus, I think we should do our own validation in any case, and if we detect issues, we should return an empty list of files to QtWebEngine after showing a warning.

AckslD commented:
@The-Compiler ah I see! Okay sounds good! Would you also say that we should validate file vs folder?

The-Compiler commented:
IMHO we should - I think it's reasonable for websites to assume that the browser will do the correct thing here, and it's reasonable for QtWebEngine to assume that the file picker will do the correct thing. We are in a bit of a special situation because we let the user configure the file picker, so I think it's fair to assume qutebrowser should validate that it's behaving as expected and not pass garbage to Qt.

AckslD commented:
@The-Compiler I now added validations and tests for these :+1:

The-Compiler commented on qutebrowser/browser/shared.py:
```diff
@@ -462,8 +472,49 @@ def _execute_fileselect_command(
+    Return:
+        List of selected files that pass the checks.
+    """
+    if not qb_mode == FileSelectionMode.multiple_files:
```
```suggestion
    if qb_mode != FileSelectionMode.multiple_files:
```

The-Compiler commented on qutebrowser/browser/shared.py:
```diff
@@ -462,8 +472,49 @@ def _execute_fileselect_command(
-    return selected_files
+            message.warning("More than one file/folder chosen, using only the first")
+            selected_files = selected_files[:1]
+    ignore_indices = set()
```
I think it might be simpler to build a new `filtered_files` list and then add them to the new list if they pass the validation, instead of saving invalid indices and then removing those indices at the end - what do you think?

You could even make this a [generator](https://book.pythontips.com/en/latest/generators.html) and use `yield selected_file` if it was valid - then you don't need a list for the end result at all. Might need an additional `list(...)` around the `_validated_selected_files` call, though.

AckslD replied:
That's indeed a lot cleaner :) Updated!

The-Compiler commented:
Thanks! Merged with some minor changes:

- 2d191f315 Break long line
- 5e4b1ffc0 Update docs
